agent/agent_endpoint.go

```diff
@@ -1283,7 +1325,12 @@ func (s *HTTPHandlers) AgentDeregisterService(resp http.ResponseWriter, req *htt
 
 func (s *HTTPHandlers) AgentServiceMaintenance(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
 	// Ensure we have a service ID
-	sid := structs.NewServiceID(strings.TrimPrefix(req.URL.Path, "/v1/agent/service/maintenance/"), nil)
+	serviceID, err := getPathSuffixUnescaped(req.URL.Path, "/v1/agent/service/maintenance/")
+	if err != nil {
+		return nil, err
```

Comment:
Why do we sometime use `return nil, err` within these functions, and other times use:
```
		fmt.Fprint(resp, "some error message")
		return nil, nil
```

?

Reply:
@kisunji, @rboyer, @kyhavlov : any ideas why this is? (I want to make sure my use of `return nil, err` is what I should be doing in these cases.)

Reply:
I raised https://github.com/hashicorp/consul/issues/11393 for this issue! We prefer `return nil, err` over writing to response.

I'm not sure about the historic context, but in go's stdlib the handlers don't return any errors and you write to `resp` directly. This pattern may have been carried over to consul's handlers.

Reply:
Linking to @Mathew-Estafanous 's https://github.com/hashicorp/consul/pull/11599 as well who started work on the refactoring

Reply:
@kisunji : thanks for creating that issue! If I understand correctly, `return nil, err` is what we want to do here and elsewhere. And applying elsewhere is captured in #11393. Yes? If so, do you want to do a final review/approve?